concordium-consensus/src/Concordium/KonsensusV1/Scheduler.hs

```diff
@@ -302,9 +308,14 @@ executeBlockPrologue BlockExecutionData{..} = do
     (mPaydayParms, theState4) <- doEpochTransition bedIsNewEpoch bedEpochDuration theState3
     -- update the seed state using the block time and block nonce
     theState5 <- doUpdateSeedStateForBlock bedTimestamp bedBlockNonce theState4
+    -- update the missed rounds count for each active baker
+    theState6 <-
+        if isJust mPaydayParms
+            then bsoClearMissedRounds theState5 activeBakers
+            else bsoUpdateMissedRounds theState5 bedMissedRounds
```

Comment:
I don't think this is how it should work.

Resetting the missed rounds on paydays is done automatically as part of `bsoRotateCurrentCapitalDistribution`.
Clearing the missed rounds for the baker should be done as part of `bsoNotifyBlockBaked`. Clearing the missed rounds for awake finalizers should be done as part of `bsoMarkFinalizationAwakeBakers`.

I think `bsoUpdateMissedRounds` should probably be called as part of `processBlockRewards`, since that is where `bsoNotifyBakedBlock` and `bsoMarkFinalizationAwakeBakers` are called. Probably, it should be ordered: mark finalization awake bakers, then update missed rounds, then notify baked block. This is because the finalization awake bakers are the ones that signed the QC for the parent block, which occurred before the missed rounds, and the new block occurred after the missed rounds. So if a validator signed the QC for the parent block, but missed a round in the interim, then it's missed round counter would be 1. Though if it also baked the new block, it would be reset to 0.

Reply:
> Resetting the missed rounds on paydays is done automatically as part of bsoRotateCurrentCapitalDistribution.
> Clearing the missed rounds for the baker should be done as part of bsoNotifyBlockBaked. Clearing the missed rounds > for awake finalizers should be done as part of bsoMarkFinalizationAwakeBakers.

I moved the clearing of missed rounds to these functions. This also has the benefit that we don't have to check or keep track of activity of a validator. I assume these functions are always called at the snapshot epoch.

Reply:
> I think bsoUpdateMissedRounds should probably be called as part of processBlockRewards, since that is where bsoNotifyBakedBlock and bsoMarkFinalizationAwakeBakers are called. Probably, it should be ordered: mark finalization awake bakers, then update missed rounds, then notify baked block. This is because the finalization awake bakers are the ones that signed the QC for the parent block, which occurred before the missed rounds, and the new block occurred after the missed rounds. So if a validator signed the QC for the parent block, but missed a round in the interim, then it's missed round counter would be 1. Though if it also baked the new block, it would be reset to 0.

Agreed! I implemented and changed the order of the operations as you describe.

Reply:
> > Resetting the missed rounds on paydays is done automatically as part of bsoRotateCurrentCapitalDistribution.
> > Clearing the missed rounds for the baker should be done as part of bsoNotifyBlockBaked. Clearing the missed rounds > for awake finalizers should be done as part of bsoMarkFinalizationAwakeBakers.
> 
> I moved the clearing of missed rounds to these functions. This also has the benefit that we don't have to check or keep track of activity of a validator. I assume these functions are always called at the snapshot epoch.

`bsoRotateCurrentCapitalDistribution` happens on the payday epoch. The others will happen every block. But - the question is, do we actually want `bsoRotateCurrentCapitalDistribution` to reset the missed round counters at all?

Reply:
> But - the question is, do we actually want bsoRotateCurrentCapitalDistribution to reset the missed round counters at all?
We should only clear it if there was activity, but I think that would already have happened with `bsoNotifyBlockBaked` or with `bsoMarkFinalizationAwakeBakers`. So bsoRotateCurrentCapitalDistribution should keep the missed rounds. Do you agree?

Reply:
This has been discussed and resolved below.